cmd/srcd/cmd/parse_test.go

```diff
@@ -0,0 +1,257 @@
+
+	out, err := s.RunCommand(context.TODO(), "parse", "drivers", "list")
+	outStr := out.String()
+
+	require.NoError(err, outStr)
+
+	/* Example output:
+
+	LANGUAGE	VERSION
+	----------	----------
+	python		v2.8.0
+	cpp		v1.1.0
+	java		v2.5.0
+	javascript	v2.6.0
+	bash		v2.4.0
+	ruby		v2.9.0
+	go		v2.5.0
+	csharp		v1.4.0
+	php		v2.7.0
+	*/
+
+	// Simple checks to see if it's the table, and contains a known driver
+	require.Contains(outStr, "LANGUAGE	VERSION")
+	expected := regexp.MustCompile(`javascript	v\S+`)
+	require.Regexp(expected, outStr)
+}
+
+func (s *ParseTestSuite) TestLang() {
+	for _, tc := range testCases {
+		s.T().Run(tc.filename, func(t *testing.T) {
+			require := require.New(t)
+
+			path := filepath.Join(s.testDir, tc.filename)
+
+			// Check the language is detected
+			out, err := s.RunCommand(context.TODO(), "parse", "lang", path)
+			require.NoError(err, out.String())
+			require.Equal(tc.lang+"\n", out.String())
+		})
+	}
+}
+
+// same as RunCommand, but captures only stdout instead of stdout + stderr
+func (s *ParseTestSuite) runCommandStdout(ctx context.Context, cmd string, args ...string) (*bytes.Buffer, error) {
+	var out bytes.Buffer
+
+	command := s.CommandContext(ctx, cmd, args...)
+	command.Stdout = &out
+
+	return &out, command.Run()
+}
+
+func (s *ParseTestSuite) TestUast() {
+	modes := []string{"default-mode", "semantic", "annotated", "native"}
+	langOpts := []string{"no-lang-option", "--lang"}
+	queryOpts := []string{"no-query-option", "--query"}
+
+	for _, tc := range testCases {
+		s.T().Run(tc.filename, func(t *testing.T) {
+			path := filepath.Join(s.testDir, tc.filename)
+
+			for _, mode := range modes {
+				t.Run(mode, func(t *testing.T) {
+
+					for _, lang := range langOpts {
+						t.Run(lang, func(t *testing.T) {
+
+							for _, query := range queryOpts {
+								t.Run(query, func(t *testing.T) {
+
+									require := require.New(t)
+
+									args := []string{"uast", path}
+
+									if mode != "default-mode" {
+										args = append(args, "--mode", mode)
+									}
+
+									if lang != "no-lang-option" {
+										args = append(args, "--lang", tc.lang)
+									}
+
+									if query != "no-query-option" {
+										// Xpath query to get the root node
+										args = append(args, "--query", "/")
+									}
+
+									// Intentionally left to help creating new test cases
+									// t.Log("about to run parse " + strings.Join(args, " "))
+
+									// RunCommand mixes stdout and stderr. To parse the UAST output properly
+									// we need to read stdout only
+									uastOut, err := s.runCommandStdout(context.TODO(), "parse", args...)
+
+									// ----------------
+									// TODO Temporary test skip, it fails for cpp, bash, and csharp.
+									// See https://github.com/src-d/engine/issues/297
+									if err != nil && strings.Contains(uastOut.String(), "is not supported") {
+										t.Skip("TEST FAILURE IS A KNOWN ISSUE (#297): " + uastOut.String())
```

Comment:
I would better `if` only for failing languages. So we would still test no-error for the rest.

Reply:
You mean like this?
```go
if err != nil &&
	(tc.lang == "c++" || tc.lang == "shell" || tc.lang == "c#") &&
	strings.Contains(uastOut.String(), "is not supported") {

	t.Skip("TEST FAILURE IS A KNOWN ISSUE (#297): " + uastOut.String())
}
```

Reply:
something like:
```go
if lang == "no-lang-option" && tc.lang == "c++" || tc.lang == "shell" || tc.lang == "c#" {
  t.Error(err) // so we know when it's fixed and can remove skip
  t.Skip("TEST FAILURE IS A KNOWN ISSUE (#297)")
}
```

the main problem currently if python, for example, returns an error it will be skipped and CI would return success.

Reply:
Changed in 7fb3fab

Reply:
better :)